kernel/sync_block/src/lib.rs

```diff
@@ -47,21 +66,55 @@ impl MutexFlavor for Block {
     where
         T: ?Sized,
     {
+        // Fast path
         // This must be a strong compare exchange, otherwise we could block ourselves
         // when the mutex is unlocked and never be unblocked.
         if let Some(guards) = Self::try_lock(mutex, data) {
-            guards
+            return guards;
+        }
+
+        // Middle path: Wait for one timeslice (see below) of the mutex holder to see if
+        // they release the lock.
+        let holder_id = data.holder.load(Ordering::Acquire);
+        if let Some(holder_task) = task::get_task(holder_id).and_then(|task| task.upgrade()) {
+            // Hypothetically, if holder_task is running on another core and is perfectly in
+            // sync with us, we would only ever check if they are running when we are also
+            // running and so we wouldn't detect when their timeslice is over. However, the
+            // likelihood of this is infinitesimally small and the code, is still correct as
+            // once the lock is released the holder will still set data.holder to 0 and we
+            // will exit the loop.
+            while holder_task.running_on_cpu().is_some()
```

Comment:
if you just want to know whether it's running, you can use this
```suggestion
            while holder_task.is_running()
```